### DISCUSSION TITLE
Getting behind the error: `The control already has a visual parent.`

### DISCUSSION THREAD

**Discussion by duck-dev** [AvaloniaUI/Avalonia#7215] [Q&A] [answered]
Hello everyone,

the following description of my issue will be quite long. You can skim all the introductions and explanations and acknowledge when the exception occurs in the heading **"Reproducing the error"** at the bottom of this post. The additional information above that will be relevant for the comprehension of my issue though. Thank you very much already in advance!

### Introduction to the issue and exception

I've stumbled across an exception that occurred after I made some changes and I'm not entirely sure what it means. I'd like to understand what exactly it means and where it occurs (the stack trace has no valuable information that affects **my** code) before I can think of a solution (which I already tried, but apparently I'm still confused).
The exception I've got is:
```
System.InvalidOperationException: The control already has a visual parent.
```
(entire stack-trace: https://paste.myst.rs/eoioofw3)
### Background information about the circumstances

The essential change, that undoubtedly caused the exception, is that I replaced "hard-coded" elements in the XAML with a `UserControl`, so I can fill different `UserControls` into it. 
This is specifically used for a system, that allows you to change the style of specific buttons, so they'll either be displayed in a grid or in a list. The thing is, that the entire layout of the button, with all it's sub-elements, changes, depending on what layout you choose. That's why I went with entirely different `UserControls`, because the structure of the sub-elements changes and simply changing/adding a class wouldn't do the job. 

So, earlier I had this for example (very simplified):
```xml
<ItemsControl Items="{Binding Items}">
    <ItemsControl.ItemsPanel>
        <ItemsPanelTemplate>
            <WrapPanel Orientation="Horizontal"/>
        </ItemsPanelTemplate>
    </ItemsControl.ItemsPanel>
    <ItemsControl.ItemTemplate>
        <DataTemplate>
            <Button Command="{Binding $parent[Window].DataContext.SomeCommand}">
                <Grid ColumnDefinitions="*,*">
                    <TextBlock Grid.Column="0" Text="Example1"/>
                    <TextBlock Grid.Column="0" Text="Example2"/>
                </Grid>
            </Button>
        </DataTemplate>
    </ItemsControl.ItemTemplate>
</ItemsControl>
```
--------------------------------------------------------------------------------------------------------------------------------------------------------------
And now I have two separate files (`GridButton.axaml` and `ListButton.axaml`) and the Button itself only contains this:
```xml
<ItemsControl Items="{Binding Items}">
    <ItemsControl.ItemsPanel>
        <ItemsPanelTemplate>
            <WrapPanel Orientation="Horizontal"/>
        </ItemsPanelTemplate>
    </ItemsControl.ItemsPanel>
    <ItemsControl.ItemTemplate>
        <DataTemplate>
            <Button Command="{Binding $parent[Window].DataContext.SomeCommand}">
                <UserControl Content="{Binding SomeBinding}" DataContext="{Binding}"/>
            </Button>
        </DataTemplate>
    </ItemsControl.ItemTemplate>
</ItemsControl>
```
The `GridButton.axaml` file would contain the actual content that needs to be filled in:
```xml
<Grid ColumnDefinitions="*,*">
    <TextBlock Grid.Column="0" Text="Example1"/>
    <TextBlock Grid.Column="0" Text="Example2"/>
</Grid>
```

`SomeBinding` in this example, would be a property of type `UserControl?` in an instance, that's included in this `ObservableCollection` that's used for the `Items` property of the `ItemsControl`. So, I create a new instance of either `GridButton` or `ListButton` and assign it to this property, so it can be used for the `UserControl` element inside the Button. I know, I could use a ViewModel and let the `ViewLocator` do it's job, but I don't want to create many ViewModels I'll never actually need, because this is a purely visual matter and I don't need any code for any of those "button-styles".

### A bit more background information about the system of switching pages

The entire system and connections between the different parts are quite complicated and there are many changes I've made to get this system to work, but I'll try to explain it well. To be able to do that, I'll need to give you, dear reader, a bit of background information about other circumstances, similar to the section above.
In the `MainWindow`, I have a `UserControl` element, that's filled with a specific `UserControl` element, which essentially lists those Buttons I'm talking about the whole time. There are 3 different kinds of views, that can be displayed, but that's not so relevant. 
The way how I change those views is quite simple:
`MainWindowViewModel.cs`:
```cs
internal void SwitchPage<T, TItems>(IEnumerable<TItems> items) where T : ListViewModelBase, IListViewModel<TItems>
{
    if (_views.Any(x => x.GetType() == typeof(T)))
    {
        if (_views.Find(x => x.GetType() == typeof(T)) is not ListViewModelBase viewModelBase)
            return;
        Content = viewModelBase;
        ((IListViewModel<TItems>)Content).Items = new ObservableCollection<TItems>(items);
    }
   else
   {
       Content = (Activator.CreateInstance(typeof(T), items) as T)!;
       _views.Add(Content);
   }
}
```
Which basically means: Find the ViewModel we need in a list. If it's there, assign it to the `Content` property. If not, create a new one of this type and assign it to the `Content` property. The `UserControl` in the `MainWindow` binds to this `Content` property, which is of type `ListViewModelBase`, a class that derives from `ViewModelBase`. I tried always instantiating a new instance and therefore not using this list, but that didn't change anything at all.

### Reproducing the error

Now, if I switch to another page, from the page the program starts from and **switch back** to this page, this exception occurs. So, **switching back to a page, that has already been opened once**, causes this exception to raise.
With "page", I mean one of those 3 possible controls/views. That's what I refer to as a page, due to it's behaviour and purpose in the application.

### What does this exception exactly mean and what would you propose?

It's quite a long description of my problem and you can probably skim some parts, but I included everything just in case, so everyone can understand the issue quite well.
I'd be very happy if someone could enlighten and help me to understand this issue better. Any suggestions are very welcome and if someone already thinks of a solution, I'd be very happy to read about it.

**Comment by danwalmsley** [comment]
Do you have a way to reproduce this? @duck-dev 

It sounds like something to do with how you are implementing your Page navigation.

Controls get attached to the visual tree when they appear on the screen.

If you try to attach them again then you get that error.

**Reply by duck-dev** [reply]
Not really I fear. I have a repository, where all those relevant changes are on the `dev` branch (which should actually be a separate branch, but I messed it up). You can take a look at that, but I guess that's not the optimal way of doing it with all those irrelevant files.
This is the repository: https://github.com/duck-dev/Grade-Management/tree/dev

Otherwise there is only the information given in my initial comment.

**Reply by duck-dev** [reply]
> It sounds like something to do with how you are implementing your Page navigation.
> 
> Controls get attached to the visual tree when they appear on the screen.
> 
> If you try to attach them again then you get that error.

Well, the Page navigation worked fine ever since. Only this `UserControl` for the button-style causes the issue right now, because **if I remove it and just "hard-code" the elements, the exception is nowhere to be seen.** This is referring to the first introduction with the `Button` element and the `UserControl` inside it. The Page navigation in the bottom part of the message has always worked fine, but is most likely indirectly connected to the issue.
It's just that I don't fully understand what this exception means. @danwalmsley

**Comment by ahopper** [comment]
Using a UserControl inside the button explicitly like this is not normal, a ContentControl might be more normal.
 Also putting UserControls inside a view model is not normal MVVM, it would be more conventional to use DataTemplates or a View locator to map viewmodel items to  views(either directly in datatemplates or in UserControls).
I realize this does not explain why you get the error but a more conventional approach might magically fix it.

**Reply by duck-dev** [reply]
@ahopper I tried it with a `ContentControl` before as well, but that didn't change anything at all. I am aware that a `ContentControl` might be "better" in certain situations, but a `UserControl` usually does the job for me.

> Also putting UserControls inside a view model is not normal MVVM, it would be more conventional to use DataTemplates or a View locator to map viewmodel items to views(either directly in datatemplates or in UserControls).

If you're referring to the `UserControl` reference (the property), the `UserControl` element in the XAML is binding to, it isn't actually inside a ViewModel. It's inside the class, whose instances are contained in the `ObservableCollection` that's used for the `ItemsControl`.

I'm very grateful for your answer, every hint is helpful, but I fear I'm still not able to get behind that issue at the moment.

**Comment by ahopper** [comment]
Keeping UserControls floating around in a collection like this sounds like asking for trouble to me.  Maybe there is a bug meaning that they don't get properly disconnected from their parent when things change or maybe you are using one twice, whatever it is using datatemplates  or a view locator does all this for you and is well proven.

**Reply by duck-dev** [reply]
Well, my thought behind that was, that I'd have to create a new instance for each of those buttons, even if it's the same type. That would logically make sense I guess. That's why I didn't keep the property in a ViewModel, but in this class. Very inconvenient, especially because I had to find a way to update the changes visually if the property gets changed. I could update the entire collection (create a new one with the same elements), but seems very inconvenient. I would be happy to find a much better way and I'm currently thinking about it.

Could you maybe elaborate on your ideas? I have something in mind, but I'm not sure if we mean the same. I'm always open to new ideas and suggestions, that's why I'm interested.
Thank you!

**Comment by ahopper** [comment]
there is an example of datatemplates here https://github.com/ahopper/zombieturkey/blob/c8c6a3c68ed221be161c8a2060ba7e9efdd70bcd/zombieturkey/Views/MainView.axaml#L39-L52
and a good description of using viewlocators here https://docs.avaloniaui.net/tutorials/todo-list-app/locating-views

**Reply by duck-dev** [reply]
Sure, I'm very familiar with the View Locator, but I mentioned above that I don't want to use View Models for that, because this is a purely visual matter and I don't need to create 6 View Models that will be empty in the end.

I'm not as familiar with DataTemplates yet. Do you suggest to create two DataTemplates inside each file, where I need them? One for the Grid-layout and one for the List-layout and then assign the respective type into the property, that's used for the Binding. That's how I've understood the usage of DataTemplates from the Avalonia documentation and that would mean that I don't need 6 views (Grid and List view for 3 pages), but only two types.

**Comment by ahopper** [comment]
you can put the datatemplates at any level in the app so they can be shared, they can also contain your user controls.

**Reply by duck-dev** [reply]
Sure. The only limitation with DataTemplates, is that it tries to use the specified type when I use a `Binding` inside it, but I actually need to use this type inside the ObservableCollection, where I can't use `$parent` either, so it's a bit of a limitation with very few ways out.
And I can't use the type of the ObservableCollection to differentiate the templates of course, because that's not intended behaviour.

I feel like I might be overcomplicating things at this point (or actually the whole time), but I've entered a loop I can't exit until I find a proper solution.

**Comment by ahopper** [comment]
you are currently putting UserControls in what is effectively a local viewmodel that you bind to. Instead of putting UserControls in there you can put some other simple object in that determines the view you want and then the datatemplate can manage the creation of the appropriate usercontrols for you.
If you want to find out why you get the original error I recommend posting a full failing project somewhere.

**Reply by duck-dev** [reply]
Yes, that's what I was trying to achieve there, but as I said above, any `Binding` inside the DataTemplate uses the specified Template type as the DataContext, but I'd need what you referred to as a "local ViewModel" (the one in the ObservableCollection, that belongs to each specific button). You can take the example in my initial post, where the Button occurs. There, I have this `ItemsControl` and inside the `DataTemplate` of an `ItemTemplate` any Binding would refer to the element from the `ObservableCollection`. That wouldn't be the case with a `DataTemplate`, where it would use the `DataType`.
Let me make an example:
```xml
<UserControl.DataTemplates>
    <DataTemplate DataType="namespace:GridButton">
        <Element Width="{Binding}"> <!-- The Binding is referring to this GridButton, instead of the context, where this Template will be inserted -->
    </DataTemplate>
</UserControl.DataTemplates>
```

**Comment by ahopper** [comment, marked as answer]
There are many ways to do this, for starters :-
The datacontext at this point (GridButton) could contain the context needed for its content.
Styles could be used as they can change the complete template of controls.
You can bind to properties of the View for some elements and properties of the viewmodel for others.
Things that at first appear to be 'View Only' often aren't e.g. users says why do I have to reset this each time I come back to this view.

**Reply by duck-dev** [reply]
Thank you very much for your answers and your patience.

I experimented with the DataTemplates and had to change quite a lot to get that to work, but it seems to work perfectly fine now, just as I imagined it. Considering the time of the day when I was wondering about the DataTemplates not being able to access the desired Binding, it's quite clear why I didn't come to that idea. I guess I shouldn't rack my brain so late about such issues. 
Now I have a reference to the respective needed instance in this `GridButton` or `ListButton` class (respectively a base class both are inheriting) and it works like a charm.

Thank you again!
